tripal/t4d8
# Content Types Creation With Terms

Issue #212

This PR updates the `ChadoPrepare` task so that it creates content types using Terms instead of only hard-coded labels.  

## Changes Added:

To make this happen I made the following changes.

1.  Added functions for importing all of the default ontologies and terms. 
  - You'll notice in the `ChadoPrepare` task that there are a lot of `addOntologyXXXX` functions. One for each ontology. Each one of these functions adds the IdSpace, Vocabulary and Terms that are required to get content types up and running. 
  -  There are more ontologies and terms than just what is needed for the content types. I actually added all of the terms used to map columns and tables of Chado in the semantic web part. But, I did not add the support for semantic web.  I just added all the terms since I was at it.
2.  Update the function for creating the entity types to be a bit more error tolerant and to automate name creation of the entity types so that it's not up to the programmer to hard-code it.
3. Minor bug fixes also included:
  - I updated the `ChadoIdSpace` and `ChadoVocabulary` classes so their error messages were a bit more informative. 
  - I fixed a bug in the `ChadoIdSpace` where it wasn't getting the default vocabulary if the IdSpace was loaded.  I could have issued a new PR for that, but it was such a small change I figured it could be looked at easy enough and this code would break if it wasn't for that fix.
  - I fixed a bug in the `ChadoIdSpace` where it now checks if the term is valid in the constructor.   It could be a perfectly good IdSpace but the is_valid flag didn't get set.

I don't think the creation of Entity Types is done, but I'm not familiar with the design to know what to do after this.  The term is being passed in when creating the entity type but I'm not sure if that's correct. So, I'm stopping the code updates until it can be looked at by @4ctrl-alt-del and @laceysanderson.   

## How to Test

1. Start with a brand new Tripal installation or drop the `chado` schema just to get a fresh start and reinstall the `tripal` and `tripal_chado` modules.
2. Install a new version of Chado
3. Run the prepare step.  You should see all of the content types get created.
